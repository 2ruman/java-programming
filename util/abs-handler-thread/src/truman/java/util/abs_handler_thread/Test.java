package truman.java.util.abs_handler_thread;

public class Test {

    private MyRequestHandler mRequestHandler;

    private final MyCallbacks mCallbacks = new MyCallbacks() {

        @Override
        public void onReceived(String message) {
            System.out.println("Handler : Received Message --> " + message);
            for (long i = 0 ; i < Integer.MAX_VALUE ; i++) {
                // Delay without sleep
            }
        }

        @Override
        public void onError(Exception e, String who) {
            System.out.println("Error: " + who);
            if (e != null) {
                System.out.println("Error Info: ");
                e.printStackTrace(System.out);
                if (e instanceof InterruptedException) {
                    System.out.println("(Not an error actually...)");
                }
            }
        }

        @Override
        public void onTerminated(String who) {
            System.out.println("Terminated: " + who);
        }
    };

    public Test test_1() {
        System.out.println("\nTest - #1: ");
        System.out.println("\tTerminating the handler in the middle of data processing\n");

        mRequestHandler = new MyRequestHandler(mCallbacks);
        mRequestHandler.start();

        MessageSender sender = new MessageSender();
        sender.start();

        try {
            Thread.sleep(3000);
        } catch (InterruptedException e) {}

        System.out.println("Terminate!");
        mRequestHandler.terminate();
        try {
            mRequestHandler.join();
        } catch (InterruptedException e) {}

        return this;
    }

    public Test test_2() {
        System.out.println("\nTest - #2: ");
        System.out.println("\tTerminating the handler while waiting\n");

        mRequestHandler = new MyRequestHandler(mCallbacks);
        mRequestHandler.start();

        mRequestHandler.add(new MyRequest(MyRequest.NONE));
        mRequestHandler.add(new MyRequest(100));
        try {
            Thread.sleep(3000);
        } catch (InterruptedException e) {}

        System.out.println("Terminate!");
        mRequestHandler.terminate();
        try {
            mRequestHandler.join();
        } catch (InterruptedException e) {}

        return this;
    }

    class MessageSender extends Thread {

        public void run() {
            System.out.println("MessageSender : Start to send messages!");
            final int totalCnt = 10;
            for (int i = 0 ; i < totalCnt ; i++) {
                MyRequest request =
                        new MyRequest(MyRequest.PRINT_MESSAGE);
                request.setData("Hello - " + i);
                mRequestHandler.add(request);
            }
            System.out.println("MessageSender : Finished! Total of the sent : " + totalCnt);
        }
    }
}
