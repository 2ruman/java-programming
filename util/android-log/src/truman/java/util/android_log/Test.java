package truman.java.util.android_log;

public class Test {

    private static final String TAG = "MyService";

    private void routine() {
        final int val = 10;

        Log.i(TAG, "routine() - Inside!");
        Log.i(null, null);

        Log.d(TAG, null);
        Log.d("ABCDEFGHIJKLMNOPQRSTU", "With long tag...");
        Log.d("CheesePizzaManagerService", "With long tag...");
        Log.d("CheesePizzaToastManagerService", "With long tag...");
        Log.d(TAG, "How much do you have?");
        Log.d(TAG, String.format("%d cents.", val));

        try {
            throw new Exception("Catch me!");
        } catch (Exception e) {
            Log.e(TAG, "Print an exception!");
            Log.printStackTrace(e);

            Log.e(TAG, "Again?");
            Log.e(e);

            Log.e(TAG, "One last time!");
            Log.e(TAG, e);
        }

        Log.i("This is for normal java code.");
        Log.d("This is for normal java code.");
        Log.e("This is for normal java code.");
        Log.d("Log file name : " + Log.getOutFileName());
    }

    /**
     * Comprehensive Testing
     */
    public Test test_1() {
        // Normal
        routine();

        // To the temporary file
        Log.setOutToFile();
        routine();

        // To the specified file with time stamp
        Log.setTimeStamp(true);
        Log.setOutToFile("my.log");
        routine();

        // Turn back to the normal
        Log.setTimeStamp(false);
        Log.setOutToFile(null);
        routine();

        return this;
    }

    /**
     *  Tag-removal Testing
     */
    public Test test_2() {
        Log.setTimeStamp(false);
        Log.setNoTag(false);

        Exception e = new Exception("Tag-removal testing");

        Log.d(TAG, "I have a tag :-p");
        Log.d(TAG, "I have a tag :-p");
        Log.d(TAG, "I have a tag :-p");
        Log.d(TAG, "I have a tag :-p");
        Log.d(TAG, "I have a tag :-p");
        Log.e(TAG, e);
        Log.setNoTag(true);
        Log.d(TAG, "Where is the tag?");
        Log.d(TAG, "Where is the tag?");
        Log.d(TAG, "Where is the tag?");
        Log.d(TAG, "Where is the tag?");
        Log.e(TAG, e);
        Log.setTimeStamp(true);
        Log.d(TAG, "Tell me where the tag is!");
        Log.d(TAG, "Tell me where the tag is!");
        Log.d(TAG, "Tell me where the tag is!");
        Log.d(TAG, "Tell me where the tag is!");
        Log.d(TAG, "Tell me where the tag is!");
        Log.e(TAG, e);

        return this;
    }

    /**
     * Multi-threading Testing
     */
    public Test test_3() {
        Log.setTimeStamp(false);
        Log.setNoTag(false);

        Thread t1 = new LoggingThread(1);
        Thread t2 = new LoggingThread(2);
        Thread t3 = new LoggingThread(3);
        Thread t4 = new LoggingThread(4);

        try {
            Log.d("==================== Ordered output ====================");
            t1.start(); t2.start();
            t1.join(); t2.join();
            Log.d("==================== Jumbled output ====================");
            t3.start(); t4.start();
            t1.join(); t2.join();
        } catch (InterruptedException e) {}

        return this;
    }

    private class LoggingThread extends Thread {
        int id;
        int repeat = 10;
        Object lock;

        LoggingThread(int id) {
            this.id = id;
            lock = (id < 3) ? Log.getLock() : new Object();
        }

        public void run() {
            synchronized (lock) {
                while (repeat-- > 0) {
                    Log.d("LoggingThread - " + id, "Hello!!!");
                    try {
                        sleep(100);
                    } catch (InterruptedException e) {}
                }
            }
        }
    }
}
