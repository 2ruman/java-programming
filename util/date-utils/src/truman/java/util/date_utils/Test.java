package truman.java.util.date_utils;

public class Test {

    public void test() {

        long time = DateUtils.getCurrentTimeMillis();
        System.out.println("Current time in ms                  : " + time);

        String formatted = DateUtils.convTimeToFormatted(time);
        System.out.println("Current time in default date format : " + formatted);

        long conv = DateUtils.convTimeToMillis(formatted);
        System.out.println("The formatted time back to ms       : " + conv);

        formatted = DateUtils.convTimeToFormatted(time, DateUtils.HYP_UDS_FORMAT);
        System.out.println("Current time with hypen/underscore  : " + formatted);

        formatted = DateUtils.convTimeToFormatted(time, DateUtils.SINGLE_SEP_FORMAT);
        System.out.println("Current time with single separator  : " + formatted);

        String wrongFormat = formatted.substring(0, 10);
        System.out.println(String.format(
                "With a wrong format(%s), it will end up with (%d)-val return after the ParseException.",
                    wrongFormat, DateUtils.convTimeToMillis(wrongFormat)));
    }
}
