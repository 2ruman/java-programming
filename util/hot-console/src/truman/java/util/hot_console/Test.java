package truman.java.util.hot_console;

public class Test {

    private static final HotConsole Out = HotConsole.getInstance();

    static {
        Out.on(true);
    }

    Thread disturber = new Thread(() -> {
        while (!Thread.currentThread().isInterrupted()) {
            Out.insertln("Hi, there!");
            Out.insertln("How do I look today?");
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    });

    public Test test_1() {
        Out.println();
        Out.println("[ How-to ]");
        Out.println(" - Type any sentence for 5 times!");
        Out.println();
        disturber.setDaemon(true);
        disturber.start();
        int i = 0;
        while (i++ < 5) {
            String s = Out.getStr();
            Out.println("#" + i + " : " + s);
        }
        disturber.interrupt();
        Out.println();
        Out.println("Bye bye!");
        return this;
    }

    public Test test_2() {
        Out.println();
        Out.println("[ How-to ]");
        Out.println(" - Press [K] for Up, [J] for Down, [H] for Left, [L] for Right");
        Out.println(" - Press [ESC] for termination...");
        Out.println();
        char c;
        while (true) {
            c = Out.getCh();
            switch (c) {
                case 'k': Out.println("Up"); break;
                case 'j': Out.println("Down"); break;
                case 'h': Out.println("Left"); break;
                case 'l': Out.println("Right"); break;
                default:
                    if (!HotConsole.isPrintable(c)) {
                        Out.println("Bye bye!");
                        return this;
                    }
            }
        }
    }

    public Test test_3() {
        Out.println();
        Out.println("[ How-to ]");
        Out.println(" - Press any of [Home], [End], [Insert], [Del], "
                + "[PageUp], [PageDown] and the arrow keys");
        Out.println(" - Press [Enter] for termination");
        Out.println();

        char c;
        SpecialKeySequence sks = null;
        while (true) {
            c = Out.getCh();
            if (sks != null) {
                sks.setCh(c);
                if (SpecialKeySequence.isKnown(sks)) {
                    SpecialKeySequence known = SpecialKeySequence.which(sks, null);
                    Out.println("[" + known.getAlias() + "]");
                    sks = null;
                } else if (sks.isFull()) {
                    sks = null;
                }
            } else if (c == 27) {
                sks = new SpecialKeySequence();
            } else if (HotConsole.isPrintable(c)) {
                Out.println("[" + c + "]");
            } else if (HotConsole.isEnter(c)) {
                Out.println("Bye bye!");
                break;
            }
        }
        return this;
    }
}
