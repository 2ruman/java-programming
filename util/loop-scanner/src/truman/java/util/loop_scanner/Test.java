package truman.java.util.loop_scanner;

public class Test {

    public void test() {

        LoopScanner scanner = new LoopScanner(
                (line) -> System.out.println("Echo: " + line));

        Thread escaper = new Thread(() -> {
            try {
                Thread.sleep(5000);
            } catch (InterruptedException e) {
            }
            System.out.println("Time to escape!");
            scanner.escape();
        });

        escaper.start();
        scanner.loop();

        System.out.println("Test finished!");
    }
}
