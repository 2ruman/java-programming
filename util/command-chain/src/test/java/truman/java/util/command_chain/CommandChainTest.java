package truman.java.util.command_chain;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

public class CommandChainTest {

    Util mUtil = new Util();

    MyCommand success_1 = new MyCommand(1, "I'm Kim");
    MyCommand success_2 = new MyCommand(2, "I'm Lee");
    MyCommand success_3 = new MyCommand(3, "I'm Choi");
    MyCommand success_4 = new MyCommand(4, "I'm Jung");
    MyCommand failure = new MyCommand(4, "I'm Park") {
        @Override
        public boolean execute() {
            super.execute();
            return false;
        }
    };

    @Test
    void testOf() {
        assertThrows(IllegalArgumentException.class, () -> CommandChain.of(null));
        AtomicReference<CommandChain> commandChain = new AtomicReference<>();
        assertDoesNotThrow(() -> commandChain.set(CommandChain.of(success_1)));
    }

    @Test
    void testLength() {
        CommandChain commandChain;

        commandChain = CommandChain.of(success_1);
        assertEquals(1, commandChain.length());

        commandChain = new CommandChain(success_1, new CommandChain(success_2, new CommandChain(success_3, new CommandChain(success_4))));
        assertEquals(4, commandChain.length());
    }

    @Test
    void testFollowedBy() {
        CommandChain commandChain = CommandChain.of(success_1)
                                                .followedBy(success_2)
                                                .followedBy(success_3)
                                                .followedBy(failure)
                                                .followedBy(success_4);
        assertEquals(5, commandChain.length());
    }

    @Test
    void testSetNext() {
        CommandChain chain1 = new CommandChain(success_1);
        CommandChain chain2 = new CommandChain(success_2);
        CommandChain chain3 = new CommandChain(success_3);
        assertEquals(1, chain1.length());
        chain1.setNext(chain2);
        assertEquals(2, chain1.length());
        chain2.setNext(chain3);
        assertEquals(3, chain1.length());
    }

    @Test
    void testHasNext() {
        CommandChain chain1 = new CommandChain(success_1);
        assertFalse(chain1.hasNext());
        chain1.setNext(new CommandChain(success_2));
        assertTrue(chain1.hasNext());
    }

    @Test
    void testTail() {
        CommandChain chain1 = new CommandChain(success_1);
        CommandChain chain2 = new CommandChain(success_2);
        CommandChain chain3 = new CommandChain(success_3);
        chain1.setNext(chain2);
        chain2.setNext(chain3);
        assertEquals(chain3, chain1.tail());
    }

    @Test
    void testSuccessCommandChain() {
        CommandChain successCommandChain = CommandChain.of(success_1)
                .followedBy(success_2).followedBy(success_3).followedBy(success_4);
        assertTrue(successCommandChain.execute());
    }

    @Test
    void approveSuccessCommandChain() {
        String actualFileName =  "CommandChainTest.approveSuccessCommandChain.received.txt";
        mUtil.runWith(actualFileName, () ->CommandChain.of(success_1)
                .followedBy(success_2).followedBy(success_3).followedBy(success_4)
                .execute());
        Approvals.verify(mUtil.getTargetFile(actualFileName));
    }

    @Test
    void testFailureCommandChain() {
        CommandChain failureCommandChain = CommandChain.of(success_1)
                .followedBy(success_2).followedBy(success_3).followedBy(failure).followedBy(success_4);
        assertFalse(failureCommandChain.execute());
    }

    @Test
    void approveFailureCommandChain() {
        String actualFileName =  "CommandChainTest.approveFailureCommandChain.received.txt";
        mUtil.runWith(actualFileName, () -> CommandChain.of(success_1)
                .followedBy(success_2).followedBy(success_3).followedBy(failure).followedBy(success_4)
                .execute());
        Approvals.verify(mUtil.getTargetFile(actualFileName));
    }

    @Test
    void testFailureCommandChainWithScheduler() {
        CommandChain failureCommandChain = CommandChain.of(success_1)
                .followedBy(success_2).followedBy(success_3).followedBy(failure).followedBy(success_4);
        ExecutorService executorService = Executors.newSingleThreadExecutor();
        Future<Boolean> future = executorService.submit(failureCommandChain);
        Boolean result = null;
        try {
            result = future.get();
        } catch (InterruptedException | ExecutionException e) {
            assertDoesNotThrow(() -> { throw e; });
        }
        executorService.shutdownNow();
        assertEquals(Boolean.FALSE, result);
    }
}