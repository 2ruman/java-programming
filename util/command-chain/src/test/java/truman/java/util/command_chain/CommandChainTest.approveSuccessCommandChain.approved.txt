execute ::MyCommand(1) - I'm Kim
execute ::MyCommand(2) - I'm Lee
execute ::MyCommand(3) - I'm Choi
execute ::MyCommand(4) - I'm Jung